### PR TITLE
Dirac cartesian basis ordering

### PR DESCRIPTION
**Pull request by camelto2**
## Proposed changes

This PR adds an additional AO basis ordering for cartesian basis functions. This follows the ordering for the DIRAC code, which will be necessary for spin-orbit testing for molecular systems. 

For reference, for cartesian gaussians the ordering is defined in Numerics/CartesianTensor.h and are hard coded in the Gamess ordering, which is 

x^L y^0 z^0
x^0 y^L z^0
x^0 y^0 z^L
x^{L-1} y^1 z^0
x^{L-1} y^0 z^1
x^1 y^{L-1} z^0
x^0 y^{L-1} z^1
...

For the DIRAC ordering, it instead uses
x^L y^0 z^0
x^{L-1} y^1 z^0
x^{L-1} y^0 z^1
x^{L-2} y^2 z^0
x^{L-2} y^1 z^1
x^{L-2} y^0 z^2
...,

This ordering is added in AOBasisBuilder::expandYlm(). Implementing the ordering here is much simpler than modifying the spinors themselves to be in gamess order. 

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
my workstation, gcc9.2

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed. 
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by rcclay** [comment]
OK to test

**Comment by ye-luo** [comment]
Please add a unit test which validate the dirac specific code paths.

File: src/QMCWaveFunctions/LCAO/AOBasisBuilder.h [hunk 1]
```diff
@@ -138,6 +139,16 @@ bool AOBasisBuilder<COT>::put(xmlNodePtr cur)
     addsignforM = 0;
   }
 
+  if (Morder == "Dirac")
+  {
+    expandlm    = DIRAC_CARTESIAN_EXPAND;
+    addsignforM = 0;
+    if (sph != "cartesian")
+    {
+      APP_ABORT(" Error: expandYlm='Dirac' only compatible with angular='cartesian'. Aborting\n");
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/AOBasisBuilder.h:148]: Please take the chance to replace all the APP_ABORT in this file.

Comment (camelto2) [reply src/QMCWaveFunctions/LCAO/AOBasisBuilder.h]: done

**Comment by camelto2** [comment]
> Please add a unit test which validate the dirac specific code paths.

Since this addition is only about the AO ordering, I added a cartesian gaussian unit test. It just creates a single orbital that is a linear combination of gaussians up to g functions, and compares the value of the orbital for both the normal gamess ordering and the new dirac ordering. 

I think this is sufficient since the gradients and laplacians of the gaussians are tested elsewhere, so this just makes sure that the ordering is correct. There is a reference python code added. 

This could also be added to in order to check the order of the spherical gaussians for various orderings as well

**Comment by ye-luo** [comment]
> > Please add a unit test which validate the dirac specific code paths.
> 
> Since this addition is only about the AO ordering, I added a cartesian gaussian unit test. It just creates a single orbital that is a linear combination of gaussians up to g functions, and compares the value of the orbital for both the normal gamess ordering and the new dirac ordering.
> 
> I think this is sufficient since the gradients and laplacians of the gaussians are tested elsewhere, so this just makes sure that the ordering is correct. There is a reference python code added.
> 
> This could also be added to in order to check the order of the spherical gaussians for various orderings as well

You did exactly what we need for a unit test.

**Pull request merged**